mir: introduce the proto-MIR

zerbina opened the pull request:
## Summary

The proto-MIR is the precursor to the MIR, acting as an intermediate
target for expressions during the AST-to-MIR translation. It is used
for the analysis previously performed on MIR expressions, allowing
`mirgen` to focus more on syntax-related translation.

Semantics don't change, and there there's no deliberate change to how
constructs are translated, although there are some minor differences
(improvements) in the produced MIR.

## Details

### Background

`mirgen` previously performed translation, semantics-related decision
making (e.g., where to inject run-time checks), and attribute
inheritance (e.g., ownership of constructed values) all in one go. For
deciding where and how to inject temporaries, the just-produced MIR was
analyzed (`isPure`, `isStable`, and `detectKind`).

While efficient in terms of memory and time usage, this approach leads
to a fairly complex `mirgen`, with the performed analysis also limited
by the MIR itself.

### The Proto-MIR

The proto-MIR (short: PMIR) is an IR separate from the MIR that only
covers a subset of existing expressions from the AST (calls,
constructions, and complex expressions, are not covered). It represents
expressions in a way such that translating them to the MIR requires as
little conditional logic as possible:
* it includes all lvalue-expression-related run-time checks
* tuple, array, and field access are separated
* `when nimvm` expressions are resolved
* purity of index and dereference operands is already ensured
* some necessary lowering is already applied

#### Data representation

* expressions are represented as *chains* of *items* (not trees of
  nodes)
* the sub-expressions forming an expression (only used for index
  operands at the moment) are part of a single `seq`
* each item stores the originating-from AST node (which still carries
  some information required by the later translation), the type, plus
  item-specific data
* all fields of `ref` type are marked as `.cursor` in order to get
  around the unnecessary ref operations (the objects are guaranteed to
  outlive the PMIR items)

#### Proto-MIR implementation

* translating expression AST to a proto-MIR expression is done via
  `exprToPmir`
* the `exprToMir` implementation is more or less a copy of
  `mirgen.genx`, but with PMIR items being emitted rather than MIR
  nodes
* the `isPure`, `isStable`, and `detectKind` (renamed to `classify`)
  analysis procedures from `mirgen` are moved to the `proto_mir` module
  and adjusted to work with the PMIR
* in contrast to the previous `isPure` implementation, the new one is
  exhaustive and also treats 'let' globals and various rvalue-yielding
  operations as pure

### Integration into `mirgen`

* whenever an expression AST is to be translated, it is first
  translated to PMIR
* the `wantX` routines are called to ensure that the expressions meet
  the contextual requirements (stable, pure, etc.)
* for expressions, `mirgen` only implements the PMIR-to-MIR step (with
  `genx`)

### Minor differences in MIR output

When field checks are disabled, a nested variant access expression such
as `a.f` is now translated to:
```
discard a.case1.case2.case3.f
```
instead of:
```
bind _1 = a.case1
bind _2 = _1.case2
discard _2.case3.f
```

This is because the operand of unchecked variant access is now never
(explicitly) captured, whereas `mirgen.genVariantAccess` previously
always captured the operand for the sake of a simpler implementation.

zerbina committed:
introduce the proto-MIR (PMIR)

It only covers simple expressions (no calls, constructions, and complex
expressions), and acts as an intermediary between the AST and the MIR,
moving most of the expression-related judgement and decision making
away from `mirgen`.

During production of the proto-MIR, there's also some lowering applied,
which is necessary for correct semantics (these lowerings were
previously applied by `mirgen`).

The procedures in the new `proto_mir` module are largely based on the
procedures from `mirgen`:
* `exprToPMir` is essentially a copy of `mirgen.genx`, but instead of
   MIR being constructed, proto-MIR is constructed
* `isStable` and `isPure` compute the same information as the
  procedures from `mirgen`, just with PMIR as the input
* `classify` is the PMIR's version of `detectKind`

zerbina committed:
mirgen: use the proto-MIR

`genx` translates proto-MIR to MIR (rather than AST to MIR) and all
callers of `genx` first translate the expression to proto-MIR. This
significantly streamlines the translation code and reduces
conditional control-flow.

zerbina committed:
proto_mir: make `knownIssue` test *fail* again

This deliberately regresses back to
`tempty_typed_expressions_issues.nim` failing, the reason being that
the underlying `sem` bug is still not fixed, and having it appear fixed
could be dangerous.

zerbina committed:
proto_mir: restore source node attachment behaviour

Both the field and the variant access MIR nodes resulting from an
`nkCheckedFieldExpr` were associated with the checked field node prior
to the introduction of the proto-MIR. This behaviour is now restored
again.

zerbina committed:
mirgen: update the module-level documentation

zerbina commented:
The name "Proto-MIR" was originally a working title. I think it fits well, but I'm also open to suggestion for better names.

saem commented on compiler/mir/proto_mir.nim:
```diff
@@ -0,0 +1,718 @@
+    pirMatLvalue ## materialize the lvalue into a temporary reference
+
+  Keep* = enum
+    kDontCare  ## can capture by value or lvalue; it doesn't matter
+    kLvalue    ## a capture must preserve lvalue-ness
+    kMutLvalue ## a capture must preserve lvalue-ness and the created
```
```suggestion
    kCaptureAny  ## can capture by value or lvalue; it doesn't matter
    kCaptureLval    ## a capture must preserve lvalue-ness
    kCaptureMutLval ## a capture must preserve lvalue-ness and the created
```

Just going based off the docs, haven't really thought about how it would read in context. This is a very soft suggestion, feel free to ignore it.

zerbina replied:
Yeah, the docs conflict with the names. My thinking was to use declarative names (i.e., what properties are expected / need to be kept) in order to make the enum more broadly applicable (in the context of the proto-MIR).

I'll update the documentation to not mention capturing.

saem commented on compiler/mir/proto_mir.nim:
```diff
@@ -0,0 +1,718 @@
+
+func isPure(e: seq[ProtoItem], n: int): bool =
+  ## Returns whether the expression at `n` is a pure expression. An expression
+  ## is pure evaluating it doesn't depend on mutable state. In other words, a
```
```suggestion
  ## is pure if evaluating it doesn't depend on mutable state. In other words, a
```

saem commented on compiler/mir/proto_mir.nim:
```diff
@@ -0,0 +1,718 @@
+  of pirMat, pirMatCursor:
+    # the materialized-into temporary is never assigned to
+    true
+  else:
```
Very soft suggestion, but perhaps a `pirNeverPureKinds` to replace the `else` and make this exhaustive is worth it?

zerbina replied:
A dedicated constant is probably a bit too much (I currently don't foresee it being used anywhere else), but the `case` should nevertheless still be exhaustive.

saem commented on compiler/mir/proto_mir.nim:
```diff
@@ -0,0 +1,718 @@
+  of pirGlobal, pirParam, pirLocal, pirConst, pirConstExpr, pirMat,
+     pirMatCursor, pirMatLvalue:
+     # location names (which includes constant expressions) and materialized
+     # expressions are never change identity
```
```suggestion
     # expressions never change identity
```

saem commented on compiler/mir/proto_mir.nim:
```diff
@@ -0,0 +1,718 @@
+    result.add ProtoItem(orig: n, typ: n.typ, kind: kind)
+    result[^1].owning = sink
+  else:
+    result.add ProtoItem(orig: n, typ: n.typ, kind: kind)
```
```suggestion
  elif kind == pirSetConstr:
    result.add ProtoItem(orig: n, typ: n.typ, kind: kind)
  else:
    result.add ProtoItem(orig: n, typ: n.typ, kind: kind)
    result[^1].owning = sink
```

Very soft suggestion, but I find avoiding negations tends to read slightly better.

saem reviewed (approved):
The comments are minor, a few typo/grammar suggestions and soft stylistic suggestions. I couldn't think of a better name than Proto-MIR.

Things are a bit cleaner with Proto-MIR, as the amount one needs to consider/understand at any given time is smaller. As alluded to in the PR message, the bigger wins are in follow-on work, so I'm hoping the analysis/codegen benefits become more pronounced then.

mrgaturus commented:
what about naming module as `mirproto.nim` to keep consistency between other mir modules naming?

zerbina commented:
> what about naming module as `mirproto.nim` to keep consistency between other mir modules naming?

My thinking was/is that the proto-MIR is not directly part of or relates to the MIR in the way the other modules do, so I deliberately not used the same nomenclature as for the other modules, in order to signify the `proto_mir` module's outlier status.

> so I'm hoping the analysis/codegen benefits become more pronounced then.

I do plan (and hope) for this to be the case. Admittedly, it took me quite a bit of effort to convince myself that introducing yet-another IR is worth it, but the two other things I experimenting with:
1. having `genx` return the information to propagate back to the caller
2. passing information about the destination's syntax and semantics expectations into `genx` and then figure out what to do within (rather than having the caller do it)

were, in my opinion, not better, as they both increased the size and complexity of the already unwieldy `mirgen`.

The immediate benefit after this PR is that it removes a large amount of duplicated decision-making from part 2 of the MIR rework. This is what motivated the proto-MIR in the first place.

zerbina committed:
proto_mir: also treat 'let' globals as pure

The same rules as for locals apply. While global 'let' bindings can
theoretically be assigned to multiple times internally during program
start-up, they're still immutable for the rest of the code.

zerbina committed:
proto_mir: clarify `Keep` documentation

Remove the "capture" wording from the comments, as `Keep` is not about
captures (it only informs how those have to work).

zerbina committed:
proto_mir: make `isPure` case statement exhaustive

Some pure rvalue operations are now also considered by the analysis. As
a consequence, `mirgen.genArgExpression` has to explicitly request an
lvalue or literal, otherwise it would receive rvalue expressions.

zerbina committed:
proto_mir: don't use negation in elif condition

zerbina committed:
fix multiple typos in comments

saem commented:
/merge
